### ISSUE TITLE
How to use local build of coreth in avalanchego?

### ISSUE THREAD

**Issue by paymog** [ava-labs/coreth#186]
I've downloaded and built avalanchego and coreth. I'm having trouble figuring out how to get avalanchego to use my local build of coreth - I've searched the avalanche docs and the docs for the two repos but haven't found instructions.

How do I get a local build of avalanchego to use a local build of coreth? Do I have to place the coreth binary under `avalanchego/build/plugins`? If so, does the binary need to have a specific name?

**Comment by aaronbuchwald** [comment]
AvalancheGo no longer uses a plugin binary to run the C-Chain VM. You can now run do this by replacing the coreth dependency with your local version. Therefore, if you have coreth and avalanchego in standard locations within your GOPATH, then you can simply go into the avalanchego directory and run the following:

```
go mod edit -replace github.com/ava-labs/coreth=../coreth && go mod tidy
```

**Comment by HanWang233** [comment]
That means if I want to run coreth based on fork repo, I only need to add
```go
replace github.com/ava-labs/coreth v0.8.5-rc.2 => github.com/<my repo>/coreth <my version>
```
in **go.mod** and run **go mod tidy**?
In your readme, coreth is a separate process, but as you describe, coreth is the package installed in avalanchego, why we need RPC to communicate?

**Comment by paymog** [comment]
@aaronbuchwald interesting - when I look at the build scripts for avalanchego it looks like a coreth binary is still used. Maybe I"m misunderstanding the build scripts?

Here are some constants that are defined like `coreth_path` and `evm_pah`: https://github.com/ava-labs/avalanchego/blob/a9c21a9c692cb6528c276ea1198009ab19db2011/scripts/constants.sh

and here's the `build_coreth` script which builds coreth and puts the binary at `evm_path`: https://github.com/ava-labs/avalanchego/blob/a9c21a9c692cb6528c276ea1198009ab19db2011/scripts/build_coreth.sh

Do you mind elaborating on what you mean by the binary plugin not being used anymore?

**Comment by paymog** [comment]
hmm, I can build `coreth` without a problem using the [build script](https://github.com/ava-labs/coreth/blob/master/scripts/build.sh) in this repo. 

```
 8:59AM /Users/paymahn/code/partyhat/coreth pending-transactions ✱ ◼
 ❯❯❯ ./scripts/build.sh
Using branch: pending-transactions
Building Coreth Version: v0.8.5; GitCommit: 36ca639ad121493fd50a27759ef684c4cf1bd44e
# gopkg.in/olebedev/go-duktape.v3
In file included from _cgo_export.c:4:
debugger.go:23:13: warning: unused function '_duk_debugger_attach' [-Wunused-function]
```

But when I use your proposed `go mod` commands I get the following errors when building:

```
 8:54AM /Users/paymahn/code/partyhat/avalanchego  ✘ 1 build.coreth
 ❯❯❯ go mod edit -replace github.com/ava-labs/coreth=../coreth && go mod tidy
 8:55AM /Users/paymahn/code/partyhat/avalanchego build.coreth ✱
 ❯❯❯ ./scripts/build.sh
Downloading dependencies...
Building AvalancheGo...
# gopkg.in/olebedev/go-duktape.v3
In file included from _cgo_export.c:4:
debugger.go:23:13: warning: unused function '_duk_debugger_attach' [-Wunused-function]
# github.com/ava-labs/coreth/plugin/evm
../coreth/plugin/evm/vm.go:91:2: cannot use &VM{} (type *VM) as type block.ChainVM in assignment:
        *VM does not implement block.ChainVM (missing Bootstrapped method)
../coreth/plugin/evm/vm.go:92:4: undefined: block.HeightIndexedChainVM
../coreth/plugin/evm/vm.go:728:10: undefined: snow.ErrUnknownState
```

Any idea what's going on here @aaronbuchwald?

**Comment by paymog** [comment]
Ah, looks like I had mismatched versions of `avalanchego` and `coreth` - I was on v1.7.4 for avalanchego and needed at least 1.7.5. However, now that I've upgraded I'm getting this build error:

```
 9:35AM /Users/paymahn/code/partyhat/avalanchego  ✘ 130 build.coreth ⬆ ⬇ ✭ ✱
 ❯❯❯ ./scripts/build.sh
Downloading dependencies...
Building AvalancheGo...
# gopkg.in/olebedev/go-duktape.v3
In file included from _cgo_export.c:4:
debugger.go:23:13: warning: unused function '_duk_debugger_attach' [-Wunused-function]
go get: downgraded github.com/ava-labs/coreth v0.8.6-rc.1 => v0.8.5-rc.2
Building Coreth @ v0.8.5-rc.2 ...
/Users/paymahn/code/partyhat/avalanchego/scripts/build_coreth.sh: line 31: cd: /Users/paymahn/go/pkg/mod/github.com/ava-labs/coreth@v0.8.5-rc.2: No such file or directory
```

It seems like the `build_coreth.sh` script isn't honoring the go mod edits. Is this script up to date and used by devs?

**Comment by paymog** [comment]
Looks like I had to make some edits to the code and run the avalanchego main file manually.

I had to add `filepath.Join(wd, "build"),` to https://github.com/ava-labs/avalanchego/blob/4b891d48d1fb9a71dd3397b73980d630a9f6227f/config/flags.go#L57-L56 so that running `go run main/main.go` would succeed.

Once that change was in place, and there existed the following directories:

```
avalanchego
├──build
          ├── avalanchego (the binary from compiling the app directory)
      	 └── plugins
	     └── evm
```

then running `go run main/main.go` would use the local checkout of coreth.
